rig-core/src/agent/prompt_request.rs

```diff
@@ -0,0 +1,231 @@
+use std::{
+    future::{Future, IntoFuture},
+    marker::PhantomData,
+};
+
+use futures::{future::BoxFuture, stream, FutureExt, StreamExt};
+
+use crate::{
+    completion::{Completion, CompletionError, CompletionModel, Message, PromptError},
+    message::{AssistantContent, UserContent},
+    tool::ToolSetError,
+    OneOrMany,
+};
+
+use super::Agent;
+
+pub trait State {}
+pub struct Simple;
+pub struct MultiTurn;
+
+impl State for Simple {}
+impl State for MultiTurn {}
```

Comment:
Wondering if the typestate pattern is really that useful, since both `SendPromptRequest` are very similar and the builder methods don't differ between both. Seems like they could be rolled into one :thinking:

Reply:
So, this typestate was much more involved since i added a `.steaming` aspect of the `PromptRequest` but removed it towards the end since I wanted to wait for the streaming rework.

I sorta agree, but I also wanted to leave the door open for `PromptRequest` to be more configurable perhaps?